### PR TITLE
Connection manager

### PR DESCRIPTION
**Pull request by losintikfos**
Enhancement to tier connection manager from `mongoengine,connection` to ensure backward compatibility and also for extendible reasons.

This enable applications using older and current version of `MongoEngine` to use mock connections, and also use new features available.

E.g.

Setting to request MongoMock instance connection:
```
        app.config['TESTING'] = True
        app.config['MONGODB_ALIAS'] = 'unittest'
        app.config['MONGODB_HOST'] = 'mongo://localhost'
```
Or use setting to instruct connection manager to provide temporary localhost instance of MongoDB connection i.e.
```
        app.config['TESTING'] = True
        app.config['TEMP_DB'] = True
```
To avoid temporary allocated localhost instance of MongoDB been deleted when test application go out of scope use the settings:
`         app.config['PRESERVE_TEMP_DB'] = true`

### CODE REVIEW

File: flask_mongoengine/connection.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+        else:
+            # Connection settings provided as a dictionary.
+            return _resolve_settings(settings, removePass)
+    else:
+        # Connection settings provided in standard format.
+        return _resolve_settings(config, removePass)
+
+def create_connection(config):
+    """
+    Connection is created base on application configuration
```
Comment (lafrech) [inline flask_mongoengine/connection.py]: s/base/based

File: flask_mongoengine/connection.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    settings. Application settings which is enabled as TESTING
+    can submit MongoMock URI or enable TEMP_DB setting to provide
+    default temporary MongoDB instance on localhost for testing
+    purposes. This connection is initiated with a separate temporary
+    directory location.
+
+    Unless PRESERVE_TEST_DB is setting is enabled in application
```
Comment (lafrech) [inline flask_mongoengine/connection.py]: s/is setting//

File: flask_mongoengine/connection.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+    configuration, temporary MongoDB instance will be deleted when
+    application instance go out of scope.
```
Comment (lafrech) [inline flask_mongoengine/connection.py]: s/go/goes/

**Comment by wojcikstefan** [comment]
This is a perfect example of a completely overblown and hacky PR that broke many things and added an extra layer of complexity only to allow using an uncommon testing strategy.

**Comment by lafrech** [comment]
@losintikfos does not seem to have much time to clean this up and among the people affected by the regressions, nobody stood up to fix this. The fact the it's a complicated piece of code doesn't make it easy and people use workarounds rather than address the root cause.

Considering 

- the lack of time / interest into fixing this
- the importance of the bugs that have been created
- the minor importance of the added feature (mongomock testing)

maybe reverting/removing this could be an option. It's never too late to get it fixed if somebody needs it and is willing to do it.

I'm not sure, however, that reverting is as easy as a `git revert`, since a few commits were merged since then.

@wojcikstefan, what do you think?

**Comment by wojcikstefan** [comment]
Agreed - I think the number of people affected by the regressions is much higher than the number of people benefitting from mongomock (with Travis/CircleCi/etc. the argument that you sometimes can't use MongoDB for your unit tests is pretty weak). Personally, we had to pin Flask-MongoEngine to v0.7.5 for our most recent project.

Yea, reverting it will probably have to go beyond `git revert`, especially given that this PR isn't properly isolated and includes many different changes (some of them desirable, many of them not). I might give it a shot sometime within in the next 2 weeks or so.

**Comment by jcass77** [comment]
Please consider #277 as a tentative attempt to get connection management working again, without making any other major changes.

**Pull request merged**